lib/spack/spack/cmd/modules/__init__.py

```diff
@@ -162,49 +163,23 @@ def loads(module_type, specs, args, out=sys.stdout):
 
 
 def find(module_type, specs, args):
-    """Returns the module file "use" name if there's a single match. Raises
-    error messages otherwise.
-    """
-
-    spec = one_spec_or_raise(specs)
-
-    if spec.package.installed_upstream:
-        module = spack.modules.common.upstream_module(spec, module_type)
-        if module:
-            print(module.path)
-        return
+    """Retrieve paths or use names of module files"""
 
-    # Check if the module file is present
-    def module_exists(spec):
-        writer = spack.modules.module_types[module_type](spec)
-        return os.path.isfile(writer.layout.filename)
+    single_spec = one_spec_or_raise(specs)
 
-    if not module_exists(spec):
-        msg = 'Even though {1} is installed, '
-        msg += 'no {0} module has been generated for it.'
-        tty.die(msg.format(module_type, spec))
-
-    # Check if we want to recurse and load all dependencies. In that case
-    # modify the list of specs adding all the dependencies in post order
     if args.recurse_dependencies:
-        specs = [
-            item for item in spec.traverse(order='post', cover='nodes')
-            if module_exists(item)
-        ]
-
-    # ... and if it is print its use name or full-path if requested
-    def module_str(specs):
-        modules = []
-        for x in specs:
-            writer = spack.modules.module_types[module_type](x)
-            if args.full_path:
-                modules.append(writer.layout.filename)
-            else:
-                modules.append(writer.layout.use_name)
-
-        return ' '.join(modules)
-
-    print(module_str(specs))
+        specs_to_retrieve = list(
+            single_spec.traverse(order='post', cover='nodes'))
```

Comment:
Shouldn't this only traverse `link` and `run` dependencies?  Those are the only ones modules would need to recursively load.  Would a user expect `spack module tcl find` to traverse `build` or `test` deps?

Reply:
The DB only stores the link/run deps, but I agree it's best to be explicit, so that is now updated.